### DISCUSSION TITLE
How to parse multipart form-data in ulfius ? In the request one is XML file and other is Json String

### DISCUSSION THREAD

**Discussion by KarurNaveen** [babelouest/ulfius#198] [Q&A] [answered]
curl -v -i -H "Content-Type:multipart/form-data" -F   "upload=@pm_omes.xml;type=application/xml" -F "data=@pm.json;type=application/json" http://localhost:8082/FaultManagerService/sendAlarms

**Comment by babelouest** [comment, marked as answer]
The [File Upload](https://github.com/babelouest/ulfius/blob/master/API.md#file-upload) paragraph in the API documentation might help you.

When a file is sent in multipart form-data, the file is available in the `request->map_post_body` as filename (key) => file_content (value), so if you send multiple files in a request, all of them will be available in the `request->map_post_body`.

If you expect large files to be uploaded, you can also use `ulfius_set_upload_file_callback_function` to intercept file uploads before they are parsed by the framework.

**Reply by KarurNaveen** [reply]
Hi @babelouest ,

curl -v -i -H "Content-Type:multipart/form-data" -F "**upload**=@pm_omes.xml;type=application/xml" -F "**data**=@pm.json;type=application/json" http://localhost:8082/OMAgentWebService/notificationFile

If i register callback ulfius_set_upload_file_callback_function , I can easily fetch the keys. (upload and data ) 
But if i dont register upload callback , but only make use of post callback , I am unable to parse the keys.

If i print the keys , i will get only **upload_filename**.

static int callback_upload_file (const struct _u_request * request, struct _u_response * response, void * user_data)
{
  char * url_params = print_map(request->map_url);
  char * headers = print_map(request->map_header);
  char * cookies = print_map(request->map_cookie);
  char * post_params = print_map(request->map_post_body);
  char* fileContent = NULL, fileName = NULL;
  const char ** keys = NULL;
  size_t len = 0 ;
        keys = u_map_enum_keys(request->map_post_body);
        for(int i=0; keys != NULL && keys[i] != NULL; i++)
        {
                printf("KEYS: %s\n ",keys[i]);
                if(!strcmp("data",keys[i])){

                        fileName = u_map_get(request->map_post_body, keys[i]);
                        char request_body[request->binary_body_length + 1];
                        o_strncpy(request_body, request->binary_body, request->binary_body_length);
       
                } else if(!strcmp("upload", keys[i])){

                        char request_body[request->binary_body_length + 1];
                        o_strncpy(request_body, request->binary_body, request->binary_body_length);


                }else {

                }
         }
}

 ulfius_add_endpoint_by_val(&(agent->rest_instance), "POST", "/OMAgentWebService/notificationFile",NULL, 1, &callback_upload_file, agent);

Thanks
Naveen

**Reply by babelouest** [reply]
In your sample code, you copy `request->binary_body`, which contains the entire request body content, so the files content and all the metadata around.
The file content is available using `u_map_get(request->map_post_body, keys[i])`:

```C
printf("filename %s\n", keys[i]);
printf("content %s\n", u_map_get(request->map_post_body, keys[i]));
```
